kong/router/fields.lua

```diff
@@ -0,0 +1,317 @@
+local buffer = require("string.buffer")
+local context = require("resty.router.context")
+
+
+local type = type
+local pairs = pairs
+local ipairs = ipairs
+local assert = assert
+
+
+-- cache key string
+local str_buf = buffer.new(64)
+
+
+local is_http = ngx.config.subsystem == "http"
+
+
+local MATCH_CTX_FUNCS
+local CACHE_KEY_FUNCS
+local get_cache_key
+
+
+if is_http then
+
+
+local tb_sort = table.sort
+local tb_concat = table.concat
+local replace_dashes_lower = require("kong.tools.string").replace_dashes_lower
+
+
+CACHE_KEY_FUNCS = {
+    ["http.method"] =
+    function(v, params, buf)
+      buf:put(params.method or ""):put("|")
+    end,
+
+    ["http.path"] =
+    function(v, params, buf)
+      buf:put(params.uri or ""):put("|")
+    end,
+
+    ["http.host"] =
+    function(v, params, buf)
+      buf:put(params.host or ""):put("|")
+    end,
+
+    ["tls.sni"] =
+    function(v, params, buf)
+      buf:put(params.sni or ""):put("|")
+    end,
+
+    ["http.headers."] =
+    function(v, params, buf)
+      local headers = params.headers
+      if not headers then
+        return
+      end
+
+      for _, name in ipairs(v) do
+        local name = replace_dashes_lower(name)
+        local value = headers[name]
+
+        if type(value) == "table" then
+          tb_sort(value)
+          value = tb_concat(value, ",")
+        end
+
+        buf:putf("%s=%s|", name, value)
+      end -- for ipairs(v)
+    end,
+
+    ["http.queries."] =
+    function(v, params, buf)
+      local queries = params.queries
+      if not queries then
+        return
+      end
+
+      for _, name in ipairs(v) do
+        local value = queries[name]
+
+        if type(value) == "table" then
+          tb_sort(value)
+          value = tb_concat(value, ",")
+        end
+
+        buf:putf("%s=%s|", name, value)
+      end -- for ipairs(v)
+    end,
+}
+
+
+MATCH_CTX_FUNCS = {
```

Comment:
Personally, I'd prefer to map from "http.method" to the params field names and make special handling of headers and queries.

Reply:
Sorry, I do not  understand what you mean, perhaps we could open another PR to improve it when this PR is merged.

Reply:
Please check the latest implement.